ocaml-lsp-server/src/custom_requests/req_merlin_call_compatible.ml

```diff
@@ -0,0 +1,156 @@
+  | `Float f -> Either.Left (string_of_float f)
+  | `Int i -> Either.Left (string_of_int i)
+  | `Intlit i -> Either.Left i
+  | x -> Either.Right x
+
+let args_from_structured_json params =
+  match List.assoc_opt "args" params with
+  | Some (`List args) -> (
+    (* Only lists containing stringish data are accepted. *)
+    match List.partition_map ~f:partition_args args with
+    | args, [] -> Some args (* No invalid arg. *)
+    | _, _ :: _ -> None (* Some invalid args. *))
+  | _ ->
+    (* If args is not a list or is absent, we assume that an empty list is
+       passed. *)
+    Some []
+
+let from_structured_json = function
+  | `Assoc params ->
+    let result_as_sexp = as_sexp_from_structured_json params in
+    let open Option.O in
+    let* command = command_from_structured_json params in
+    let* args = args_from_structured_json params in
+    let* uri = uri_from_structured_json params in
+    Some { result_as_sexp; command; args; uri }
+  | _ -> None
+
+let raise_invalid_params ?data ~message () =
+  let open Jsonrpc.Response.Error in
+  raise @@ make ?data ~code:Code.InvalidParams ~message ()
+
+let from_structured_json_exn = function
+  | None -> raise_invalid_params ~message:"Expected params but received none" ()
+  | Some params -> (
+    match from_structured_json params with
+    | Some params -> params
+    | None ->
+      let data =
+        `Assoc
+          [ ("expectedParams", expected_params)
+          ; ("receivedParams", (params :> Json.t))
+          ]
+      in
+      raise_invalid_params ~data ~message:"Unexpected params format" ())
+
+let with_pipeline state uri f =
+  match Document_store.get_opt state.State.store uri with
+  | Some doc ->
+    let merlin = Document.merlin_exn doc in
+    Document.Merlin.with_pipeline_exn merlin f
+  | None ->
+    let message =
+      "Document " ^ Uri.to_string uri ^ " wasn't found in the document store"
+    in
+    raise_invalid_params ~message ()
+
+let parse_arguments state uri specs raw_args cmd_args =
+  let conf = Merlin_config.DB.get state.State.merlin_config uri in
+  let open Fiber.O in
+  let+ config = Merlin_config.config conf in
+  let specs = List.map ~f:snd specs in
+  Mconfig.parse_arguments
+    ~wd:(Sys.getcwd ())
+    ~warning:ignore
+    specs
+    raw_args
+    config
+    cmd_args
+  |> snd
+
+let perform_query action params pipeline =
+  let action () = action pipeline params in
+  let class_, output =
+    match action () with
+    | result -> ("return", result)
+    | exception Failure message -> ("failure", `String message)
+    | exception exn ->
+      let message = Printexc.to_string exn in
+      ("exception", `String message)
+  in
+  `Assoc [ ("class", `String class_); ("value", output) ]
+
+let on_request ~params state =
+  Fiber.of_thunk (fun () ->
+      let { result_as_sexp; command; args; uri } =
+        from_structured_json_exn params
+      in
+      match
+        Merlin_commands.New_commands.(find_command command all_commands)
```

Comment:
Given that this API is new, can we make it not raise an exception?

Reply:
No change is worth a re-release of any existing version in my eyes. However, a new version that contains the improved API is maybe a possibility?

As a side note, It feels wrong to introduce a suboptimal API and refuse to improve it before it even has any users. Are we waiting so it to gain some users before we make a change so that we can break the API for them? Seems counter-intuitive.

This particular issue isn't very important, but I do think it points out a more general flaw of how development is handled. There should be a tighter feedback loop between how the merlin library is developed and its only client (for the forseable future) so that we realize when an API is suboptimal ASAP and not after it's too late to change it.

Reply:
It's true that we still have a somewhat slow release cycle for Merlin, but we are in no way opposed to make quick evolution to the API when needed. We are hopefully leaning toward the tighter feedback loop you're mentioning now that we are getting more hands-on with ocaml-lsp dev. 

This particular function is so ad-hoc in its usage by ocaml-lsp, and exposing a deep merlin-specific behavior, that it did not seems worthy to rework its API when exposing it. It is very unlikely that this part of the API will ever change, unless we stop supporting the legacy Merlin protocol altogether. 

Still, since we have releases planned soon anyway, I don't mind adding an option or result-based version of that function to the API, which we will also backport to older versions of merlin-lib so that this PR can be merged.